### ISSUE TITLE
phastSim 0.0.3 crashed when simulating Indels

### ISSUE THREAD

**Issue by trongnhanuit** [NicolaDM/phastSim#16]
Dear authors,
When simulating Indels, I found that phastSim 0.0.3 crashed in some of my tests as the following. Could you please help me to have a look at it? Thanks in advance.

**- Input files:**
+ tree100.nwk
`((((((T1:0.0230671818,((T22:0.0199671195,T61:0.0297059234):0.1030019497,(T28:0.1354795694,((T46:0.1555897146,T95:0.2407945609):0.0767870727,(T52:0.0220646671,T90:0.1546463113):0.0418550348):0.0420071260):0.0332679438):0.0019240000):0.0599656837,(T13:0.0465215113,T58:0.1058430499):0.0556869562):0.0767870727,((((T5:0.0011480000,T99:0.0384192973):0.2688247574,T81:0.0309246250):0.0238257189,(T7:0.0646263595,((T9:0.0455706325,((T82:0.2162823151,T93:0.0106472245):0.5298317367,T88:0.0553385238):0.0597837001):0.1294627173,T70:0.0260066905):0.0677273831):0.0160168752):0.0596020470,(((T6:0.0579818495,T21:0.0798507696):0.1142564176,((T12:0.1070024832,T45:0.0850971266):0.2918771232,((T35:0.1795767491,(T77:0.0011170000,T85:0.0848632083):0.0299754654):0.0092115289,(T66:0.0336872317,T86:0.2631089160):0.3270169119):0.0642454066):0.0106472245):0.0843970070,((T11:0.1090644119,T78:0.2234926445):0.0099820335,T31:0.0061875404):0.0286349627):0.2900422094):0.0103140759):0.0450985623,(T2:0.0544727175,((T8:0.0616186139,(T57:0.1439695138,T62:0.1072944542):0.0759286983):0.1370421012,(T38:0.0406465608,T48:0.0141563564):0.2282782466):0.0146182510):0.1692819521):0.2353878387,((((((T3:0.1527857925,T42:0.1532476871):0.1386294361,(T19:0.0848632083,((T26:0.1207311706,T84:0.1331806176):0.0318828801,((T37:0.1237874356,T96:0.0430782916):0.1410587054,((T49:0.0444725822,T68:0.0019270000):0.0121038328,T64:0.2103734234):0.0570929548):0.0867500568):0.0471604911):0.0650087691):0.0365283318,(((((T15:0.0317454231,(T76:0.0677273831,T89:0.4422848629):0.0320205264):0.0381260419,T72:0.2189256408):0.0740238788,(T33:0.0031490667,(T69:0.0154317360,T94:0.2071473372):0.2312635429):0.0657780037):0.0279713903,((T32:0.0519193873,T87:0.0170788321):0.0650087691,T36:0.0772190388):0.1231001477):0.0116533816,((((T24:0.1687399454,T73:0.0359536176):0.0497580397,(T43:0.0634878272,T67:0.3057607677):0.0146182510):0.0502526821,T41:0.2703062660):0.1518683549,((T25:0.0077961541,T91:0.0715392790):0.2563949857,T51:0.0567395975):0.2918771232):0.0099820335):0.3540459449):0.2407945609,((((T14:0.0619896719,T53:0.3912023005):0.1897119985,((T40:0.0353821875,(T83:0.1660731207,T97:0.0570929548):0.0083381609):0.0285018955,(T50:0.1666008264,T71:0.1130102956):0.1877317358):0.0294371061):0.2180367460,((T23:0.0921303274,T92:0.0776528789):0.0088831214,T29:0.1422958345):0.0032523192):0.0258770729,T20:0.0874669057):0.0132389188):0.3296837366,(((T10:0.0407968238,T54:0.1650259907):0.0281037530,(T47:0.0429245637,T56:0.2128631786):0.0327116142):0.0823255866,T39:0.1374365790):0.2145581344):0.0701179352,(((T4:0.1931021537,(T63:0.1093624747,T98:0.3218875825):0.0816445397):0.1496109227,T60:0.5298317367):0.0013680000,((T16:0.0478035801,((T18:0.0738144546,(T59:0.2748872196,T74:0.0256183405):0.1016111067):0.0906340401,(T27:0.3015934981,(T44:0.0199671195,T75:0.1579879110):0.1120857898):0.1187443502):0.1237874356):0.0197232170,T17:0.1851509474):0.1645065090):0.0999672341):0.1220779923):0.2253794929,T0:0.0733969175,((T30:0.0894040123,((T65:0.0757152511,T80:0.6907755279):0.1742969305,T79:0.1589635285):0.0410980289):0.0022245609,(T34:0.1064210862,T55:0.0023268627):0.0260066905):0.0047091608);`
+ [MN908947.3.fasta](https://raw.githubusercontent.com/NicolaDM/phastSim/main/phastSim/example/MN908947.3.fasta)

**- Command:**
`phastSim --outpath .  --createPhylip --seed 1 --treeFile tree100.nwk --reference MN908947.3.fasta --indels --insertionRate CONSTANT 0.1 --deletionRate CONSTANT 0.1 --insertionLength NEGBINOMIAL 1 0.2 --deletionLength NEGBINOMIAL 1 0.2 --outputFile output`

**- Error:**
```
Traceback (most recent call last):
  File "./phastSim", line 213, in <module>
    genome_tree.write_genome_short(
  File "../.local/lib/python3.9/site-packages/phastSim/phastSim.py", line 1847, in write_genome_short
    self.writeGenomeShortIndels(
  File "../.local/lib/python3.9/site-packages/phastSim/phastSim.py", line 1698, in wrappedFunction
    writingFunction(self, node, file, mutDict, insertionDict, **kwargs)
  File "../.local/lib/python3.9/site-packages/phastSim/phastSim.py", line 1803, in writeGenomeShortIndels
    self.writeGenomeShortIndels(c, file, mutDict, insertionDict, alternative_output_format=alternative_output_format)
  File "../.local/lib/python3.9/site-packages/phastSim/phastSim.py", line 1698, in wrappedFunction
    writingFunction(self, node, file, mutDict, insertionDict, **kwargs)
  File "../.local/lib/python3.9/site-packages/phastSim/phastSim.py", line 1803, in writeGenomeShortIndels
    self.writeGenomeShortIndels(c, file, mutDict, insertionDict, alternative_output_format=alternative_output_format)
  File "../.local/lib/python3.9/site-packages/phastSim/phastSim.py", line 1698, in wrappedFunction
    writingFunction(self, node, file, mutDict, insertionDict, **kwargs)
  File "../.local/lib/python3.9/site-packages/phastSim/phastSim.py", line 1803, in writeGenomeShortIndels
    self.writeGenomeShortIndels(c, file, mutDict, insertionDict, alternative_output_format=alternative_output_format)
  File "../.local/lib/python3.9/site-packages/phastSim/phastSim.py", line 1698, in wrappedFunction
    writingFunction(self, node, file, mutDict, insertionDict, **kwargs)
  File "../.local/lib/python3.9/site-packages/phastSim/phastSim.py", line 1803, in writeGenomeShortIndels
    self.writeGenomeShortIndels(c, file, mutDict, insertionDict, alternative_output_format=alternative_output_format)
  File "../.local/lib/python3.9/site-packages/phastSim/phastSim.py", line 1698, in wrappedFunction
    writingFunction(self, node, file, mutDict, insertionDict, **kwargs)
  File "../.local/lib/python3.9/site-packages/phastSim/phastSim.py", line 1803, in writeGenomeShortIndels
    self.writeGenomeShortIndels(c, file, mutDict, insertionDict, alternative_output_format=alternative_output_format)
  File "../.local/lib/python3.9/site-packages/phastSim/phastSim.py", line 1698, in wrappedFunction
    writingFunction(self, node, file, mutDict, insertionDict, **kwargs)
  File "../.local/lib/python3.9/site-packages/phastSim/phastSim.py", line 1803, in writeGenomeShortIndels
    self.writeGenomeShortIndels(c, file, mutDict, insertionDict, alternative_output_format=alternative_output_format)
  File "../.local/lib/python3.9/site-packages/phastSim/phastSim.py", line 1698, in wrappedFunction
    writingFunction(self, node, file, mutDict, insertionDict, **kwargs)
  File "../.local/lib/python3.9/site-packages/phastSim/phastSim.py", line 1803, in writeGenomeShortIndels
    self.writeGenomeShortIndels(c, file, mutDict, insertionDict, alternative_output_format=alternative_output_format)
  File "../.local/lib/python3.9/site-packages/phastSim/phastSim.py", line 1698, in wrappedFunction
    writingFunction(self, node, file, mutDict, insertionDict, **kwargs)
  File "../.local/lib/python3.9/site-packages/phastSim/phastSim.py", line 1803, in writeGenomeShortIndels
    self.writeGenomeShortIndels(c, file, mutDict, insertionDict, alternative_output_format=alternative_output_format)
  File "../.local/lib/python3.9/site-packages/phastSim/phastSim.py", line 1698, in wrappedFunction
    writingFunction(self, node, file, mutDict, insertionDict, **kwargs)
  File "../.local/lib/python3.9/site-packages/phastSim/phastSim.py", line 1803, in writeGenomeShortIndels
    self.writeGenomeShortIndels(c, file, mutDict, insertionDict, alternative_output_format=alternative_output_format)
  File "../.local/lib/python3.9/site-packages/phastSim/phastSim.py", line 1698, in wrappedFunction
    writingFunction(self, node, file, mutDict, insertionDict, **kwargs)
  File "../.local/lib/python3.9/site-packages/phastSim/phastSim.py", line 1803, in writeGenomeShortIndels
    self.writeGenomeShortIndels(c, file, mutDict, insertionDict, alternative_output_format=alternative_output_format)
  File "../.local/lib/python3.9/site-packages/phastSim/phastSim.py", line 1746, in wrappedFunction
    if mutDict[refPos][2][offset] == m.deletedPositions[deletedChars]:
KeyError: 29487
```

Could someone help me to check it? Many thanks.
Cheers,
Nhan

**Comment by willboulton** [comment]
Dear Nhan,
Thank you very much for sending this issue - I'll try to resolve it as soon
as possible!

The other issue you posted is I believe a versioning issue as we had
spotted this as well and had a fix on github - I suspect this means an
incorrect version was pushed to pypi, I'll try to resolve this too.

Best wishes, and thanks for letting us know,
Will

**Comment by willboulton** [comment]
Dear Nhan,

Would it be possible in the first instance for you to try running `pip
install --upgrade phastSim` ? This should upgrade you to phastSim 0.0.4
which is the latest version, and I believe at the very least the
--rootGenomeLength issue should be resolved for you.

I will investigate the issue that you've mentioned here too - there is a
chance that has also been resolved since version 0.0.3 as well,

Best wishes,
Will

**Comment by willboulton** [comment]
Dear Nhan, 

Would it be possible in the first instance for you to try running `pip install --upgrade phastSim` (or `pip uninstall phastSim` and then `pip install phastSim`) ? This should upgrade you to phastSim 0.0.4 which is the latest version, and I believe at the very least the --rootGenomeLength issue should be resolved for you.

I will investigate the issue that you've mentioned here too - I believe this should have been resolved in version 0.0.4 as well. 

By the way I noticed in the command you sent that you're using both --createPhylip and --indels, however this would print an error saying "Phylip output is not yet supported with indels" - so a better option would be --createFasta. I will make this clearer in the README as currently the only mention of this is: "Output can be created in FASTA format (`--createFasta`) and/or PHYLIP format (`--createPhylip`); in the case of simulations with indels, the FASTA and PHYLIP formats will contain unaligned sequences."

Finally, using --outputFolder as . would end up prepending a . to your output files, making them hidden. This is a bug and I will resolve it since it is not the behaviour you would expect! In the short term though, the way around this would be to use --outputFolder ./ and --outputFile my_test (or you can leave it, the default is sars-cov-2_simulation_output, and it would overwrite any other output files written in the same location with the same name). 

Best wishes,
Will

**Comment by trongnhanuit** [comment]
Dear @willboulton,
I can confirm that this issue was fixed on version 0.0.4. I've just tested it. 
Thank you very much for your information/support.

Best regards,
Nhan

**Comment by willboulton** [comment]
Great! I'll leave this issue open though in case you run into any more problems,
Best wishes, 
Will
